src/collapse/dtrace.rs

```diff
@@ -0,0 +1,219 @@
+use super::Collapse;
+use smallvec::SmallVec;
+use std::collections::{HashMap, VecDeque};
+use std::io;
+use std::io::prelude::*;
+
+/// Settings that change how frames are named from the incoming stack traces.
+///
+/// All options default to off.
+#[derive(Clone, Debug, Default)]
+pub struct Options {
+    /// Include raw addresses (e.g., `0xbfff0836`) where symbols can't be found.
+    pub include_addrs: bool,
+
+    /// Annotate kernel functions with a `_[k]` suffix.
+    pub annotate_kernel: bool,
+}
+
+/// A stack collapser for the output of `perf script`.
+///
+/// To construct one, either use `perf::Folder::default()` or create an [`Options`] and use
+/// `perf::Folder::from(options)`.
+#[derive(Default)]
+pub struct Folder {
+    /// All lines until the next empty line are stack lines.
+    in_event: bool,
+
+    /// Skip all stack lines in this event.
+    skip_stack: bool,
+
+    /// Function entries on the stack in this entry thus far.
+    stack: VecDeque<String>,
+    count: usize,
+    /// Number of times each call stack has been seen.
+    occurrences: HashMap<String, usize>,
+
+    /// Current comm name.
+    ///
+    /// Called pname after original stackcollapse-perf source.
+    pname: Option<String>,
+
+    //    event_filtering: EventFilterState,
+    opt: Options,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+        //TODO: clean this up, we skip the first 4 lines:
+        //        CPU     ID                    FUNCTION:NAME
+        reader.read_line(&mut line)?;
+        //        0  64091                        :tick-60s
+        reader.read_line(&mut line)?;
+        //
+        reader.read_line(&mut line)?;
+        //
+        reader.read_line(&mut line)?;
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            if line.starts_with('#') {
+                continue;
+            }
+
+            let line = line.trim_end();
+            if line.is_empty() {
+                self.after_event();
+            } else {
+                self.on_line(line);
+            }
+        }
+        self.after_event();
+        self.finish(writer)
+    }
+}
+
+impl From<Options> for Folder {
+    fn from(mut opt: Options) -> Self {
+        Self {
+            in_event: false,
+            skip_stack: false,
+            stack: VecDeque::default(),
+            count: 0,
+            occurrences: HashMap::default(),
+            pname: None,
+            opt,
+        }
+    }
+}
+
+impl Folder {
+    fn on_line(&mut self, line: &str) {
+        self.in_event = true;
+        self.on_stack_line(line)
+    }
+
+    fn stack_line_parts(line: &str) -> Option<(&str, &str)> {
+        let mut line = line.splitn(2, '+');
```

Comment:
I think you want `rsplitn` here, since we want the [last](https://github.com/brendangregg/FlameGraph/blob/9d7c1b1233839de1f842135be84e54f2470adb42/stackcollapse.pl#L74) `+`.

Reply:
👍